components/tidb_query/src/rpn_expr/impl_string.rs

```diff
@@ -931,6 +953,75 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_upper() {
+        // Test non-binary string case
+        let cases = vec![
+            (Some(b"hello".to_vec()), Some(b"HELLO".to_vec())),
+            (Some(b"123".to_vec()), Some(b"123".to_vec())),
+            (
+                Some("café".as_bytes().to_vec()),
+                Some("CAFÉ".as_bytes().to_vec()),
+            ),
+            (
+                Some("数据库".as_bytes().to_vec()),
+                Some("数据库".as_bytes().to_vec()),
+            ),
+            (
+                Some("ночь на окраине москвы".as_bytes().to_vec()),
+                Some("НОЧЬ НА ОКРАИНЕ МОСКВЫ".as_bytes().to_vec()),
+            ),
+            (
+                Some("قاعدة البيانات".as_bytes().to_vec()),
+                Some("قاعدة البيانات".as_bytes().to_vec()),
+            ),
+            (None, None),
+        ];
+
+        for (arg, exp) in cases {
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(arg)
+                .evaluate(ScalarFuncSig::Upper)
+                .unwrap();
+            assert_eq!(output, exp);
```

Comment:
You can merge the binary / non-binary cases together. For binary, you only need to assert that the output is the same as input.

Reply:
After separating signatures, we have `ScalarFuncSig::Upper` and `ScalarFuncSig::UpperBinary`, and we don't need to merge the binary / non-binary cases together.

Reply:
After separating signatures they need to stay in different test functions as well..